app/Platform/Actions/ResolveAchievementSetsAction.php

```diff
@@ -0,0 +1,201 @@
+     *
+     * @return Collection<int, GameAchievementSet>
+     */
+    private function getAllAchievementSets(int $gameId, ?array $types): Collection
+    {
+        $query = GameAchievementSet::with([
+            'achievementSet' => [
+                'achievements' => fn ($q) => $q->orderBy('DisplayOrder'),
+                'incompatibleGameHashes',
+            ],
+        ])->where('game_id', $gameId);
+
+        if ($types !== null) {
+            $query->whereIn('type', $types);
+        }
+
+        return $query->get();
+    }
+
+    /**
+     * Fetches user preferences for the given achievement sets.
+     *
+     * @param Collection<int, GameAchievementSet> $allSets
+     * @return Collection<int, UserGameAchievementSetPreference>
+     */
+    private function getUserSetPreferences(User $user, Collection $allSets): Collection
+    {
+        return UserGameAchievementSetPreference::where('user_id', $user->id)
+            ->whereIn('game_achievement_set_id', $allSets->pluck('id'))
+            ->get()
+            ->keyBy('game_achievement_set_id');
+    }
+
+    /**
+     * Filter the achievement sets based on user preferences and hash compatibility.
+     *
+     * @param Collection<int, GameAchievementSet> $allSets
+     * @param Collection<int, UserGameAchievementSetPreference> $userSetPreferences
+     * @return Collection<int, GameAchievementSet>
+     */
+    private function filterSets(
+        Collection $allSets,
+        GameHash $gameHash,
+        User $user,
+        Collection $userSetPreferences
+    ): Collection {
+        return $allSets->filter(function (GameAchievementSet $set) use ($gameHash, $user, $userSetPreferences) {
+            // Exclude if this hash is marked as incompatible
+            if ($set->achievementSet->incompatibleGameHashes->contains('id', $gameHash->id)) {
+                return false;
+            }
+
+            // Check user-specific preferences
+            $preference = $userSetPreferences->get($set->id);
+            if ($preference !== null) {
+                return $preference->opted_in;
+            }
+
+            // Apply global preference for non-core sets
+            if ($set->type !== AchievementSetType::Core) {
+                return !$user->is_globally_opted_out_of_subsets;
+            }
+
+            // Include core sets by default
+            return true;
+        });
+    }
+
+    /**
+     * Sort the achievement sets based on their type priority.
+     *
+     * @param Collection<int, GameAchievementSet> $sets
+     * @return Collection<int, GameAchievementSet>
+     */
+    private function sortSets(Collection $sets): Collection
+    {
+        return $sets->sortBy(function (GameAchievementSet $set) {
+            return match ($set->type) {
+                AchievementSetType::Exclusive => 0,
+                AchievementSetType::Core => 1,
+                AchievementSetType::Bonus => 2,
+                AchievementSetType::Specialty => 3,
```

Comment:
I wonder if specialty should be above bonus. If I'm playing a low-level specialty set, I might care more about it that the bonus sets.

Reply:
Happy to make this change, updated in latest.